### ISSUE TITLE
Components with type parameters

### ISSUE THREAD

**Issue by ziggystar** [shadaj/slinky#197]
I try to write a Component that renders a table from a list of arbitrary values:

```
case class TableViewer[A](columns: Seq[(String, A => ReactElement)]) extends StatelessComponentWrapper {
  case class Props(rows: Seq[A])

  class Def(jsProps: js.Object) extends Definition(jsProps) {
    override def render(): ReactElement = table(className := "TableViewer")(
      tr(columns.map(_._1).map(th(_))) +:
        props.rows.map(row => tr(columns.map(_._2).map(view => td(view(row)))))
    )
  }

  def apply(rows: Seq[A]): KeyAndRefAddingStage[Def] = this.apply(Props(rows))
}
```

Then I try to use it like this:

    TableViewer[String](Seq("id" -> identity[String]))(Seq("a","b"))

But this doesn't work. The complete page stays empty, and in the JS console, I get the following error:

>uncaught exception: scala.scalajs.runtime.UndefinedBehaviorError: An undefined behavior was detected: [object Object] is not an instance of <snip>.TableViewer

**Comment by shadaj** [comment]
@ziggystar generally, `ComponentWrapper`s must be objects, you're likely to run into issues if defining it as a class. Would it be possible for you to move the `columns` into the `Props`? Then the component should work without a problem.

**Comment by ziggystar** [comment]
@shadaj That's my current solution, and it also involves casting. I would like to get rid of casting.

I tried to put a type parameter on `Props`, but then I got the error `class Props takes type parameters`.
```

@react class TableViewerTyped extends StatelessComponent {
  case class Props[A](columns: Seq[(String, A => ReactElement)], rows: Seq[A], rowClass: A => Seq[String]){
    def render(): ReactElement = table(className := "TableViewer")(
      thead(tr(columns.map(_._1).map(th(_)))),
      tbody(
        rows.map(row => tr(
          className := rowClass(row).mkString(" ")
        )(
          columns.map(_._2).map(view => td(view(row)))
        ))
      )
    )
  }

  override def render(): ReactElement = props.render()
}

object TableView{
  def applyTyped[A](columns: Seq[(String, A => ReactElement)], rowClasses: A => Seq[String] = (_: A) => Seq.empty[String])(rows: Seq[A]): ReactElement =
    TableViewerTyped(columns, rows, rowClasses)
}
```

**Comment by a1russell** [comment]
@shadaj Would it be possible to add support to the annotation for type parameters? I am also trying to this "Props takes type parameters" error, and I have seen it reported previously, as well (#40). Having to rely on casting defeats the purpose of using Scala, and causes some headache to figure out what the problem is. I don't think patterns like HOC are uncommon in the React world, so I hope it's not an unreasonable feature to request.

**Comment by shadaj** [comment]
Note to self: this situation can be improved in the short term by making hot-reloading fallback when the `Props` type is not defined inside an object. This will enable *classes* to extend `ComponentWrapper`, and then the wrapper can just take type parameters.

**Comment by shadaj** [comment]
Fixed by #287
